modules/sdk-core/src/bitgo/wallet/wallet.ts

```diff
@@ -1463,14 +1465,111 @@ export class Wallet implements IWallet {
+          throw new Error('Keychain parameters require toPubKey, and path.');
+        }
+      }
+
+      const needsKeychain = shareOption.permissions && shareOption.permissions.indexOf('spend') !== -1; // Remove !shareOption.keychain
+      let sharedKeychain: any = undefined;
+
+      if (needsKeychain) {
+        try {
+          const keychain = await this.getEncryptedUserKeychain();
+
+          if (!keychain.encryptedPrv) {
+            sharedKeychain = undefined;
+          } else {
+            const userPrv = decryptKeychainPrivateKey(this.bitgo, keychain, params.walletPassphrase);
+            if (!userPrv) {
+              throw new Error('Unable to decrypt user keychain.');
+            }
+
+            const eckey = makeRandomKey();
+            const sharing = shareOption;
+            assert(sharing.keychain?.toPubKey);
+
+            const secret = getSharedSecret(eckey, Buffer.from(sharing.keychain?.toPubKey, 'hex')).toString('hex');
+            const newEncryptedPrv = this.bitgo.encrypt({ password: secret, input: userPrv });
+
+            let pub = keychain.pub ?? keychain.commonPub;
+            if (keychain.commonKeychain) {
+              pub =
+                this.baseCoin.getMPCAlgorithm() === 'eddsa'
+                  ? EddsaUtils.getPublicKeyFromCommonKeychain(keychain.commonKeychain)
+                  : EcdsaUtils.getPublicKeyFromCommonKeychain(keychain.commonKeychain);
+            }
+
+            sharedKeychain = {
+              pub,
+              encryptedPrv: newEncryptedPrv,
+              fromPubKey: eckey.publicKey.toString('hex'),
+              toPubKey: sharing.keychain?.toPubKey,
+              path: sharing.keychain?.path,
+            };
+          }
+        } catch (e) {
+          if (e.message === 'No encrypted keychains on this wallet.') {
+            sharedKeychain = {};
+          } else {
+            throw e;
+          }
+        }
+        bulkCreateShareOptions.shareOptions.push({
+          user: shareOption.user,
+          permissions: shareOption.permissions,
+          keychain: Object.keys(sharedKeychain ?? {}).length === 0 ? undefined : sharedKeychain,
+        });
+      }
+    }
+    return await this.createBulkShares(bulkCreateShareOptions);
+  }
+
+  /**
+   * Send an encrypted wallet share to BitGo.
+   * @param params
+   */
+  async createBulkShares(params: BulkCreateShareOptions = { shareOptions: [] }): Promise<any> {
+    if (!params.shareOptions || Object.keys(params.shareOptions).length === 0) {
+      throw new Error('shareOptions cannot be empty');
+    }
+    params.shareOptions.forEach((shareOption) => {
+      if (shareOption.keychain) {
+        const { pub, encryptedPrv, fromPubKey, toPubKey, path } = shareOption.keychain;
+        if (!pub || !encryptedPrv || !fromPubKey || !toPubKey || !path) {
+          throw new Error('requires keychain parameters - pub, encryptedPrv, fromPubKey, toPubKey, path');
+        }
+      }
+    });
```

Comment:
you can use discriminated union types for your params so you wouldn't need to have this check i think.

Reply:
Actually I was following this format - https://github.com/BitGo/BitGoJS/blob/master/modules/sdk-core/src/bitgo/wallet/wallet.ts#L1452-L1460
Its also easier for the caller to understand the error if the caller has missing keychain parameter.
Should I still modify the code to use `discriminated union types`??

Reply:
yes if it's not a huge refactor, because if you're using a good union, the caller would get an error during compilation and not run time.

Reply:
Hi @mohammadalfaiyazbitgo 
Modified, the distributed union types is being enforced by - https://github.com/BitGo/BitGoJS/pull/4884/files#diff-d998f4b5c3a58f8f1ac3c8075b65be2eef631b7e951bb9af3fa81ccf4af52642R492
We can remove the 
`if (!pub || !encryptedPrv || !fromPubKey || !toPubKey || !path) {
          throw new Error('requires keychain parameters - pub, encryptedPrv, fromPubKey, toPubKey, path');
        }` it will give compile time error - `No match for request`. 

However, by using this check, we can provide the caller with better error logging. 
Should I still remove the check?